application/ledger_query_service/src/bitcoin/block_processor.rs

```diff
@@ -1,440 +1,70 @@
-        for (query_id, query) in block_queries.all() {
-            trace!("Matching query {:#?} against block {:#?}", query, block);
-            let block_id = block_id.clone();
-            let result_future = query.matches(block).map(move |result| match result {
-                QueryMatchResult::Yes { .. } => {
-                    trace!("Block {} matches Query-ID: {:?}", block_id, query_id);
-                    Some((query_id, block_id))
-                }
-                QueryMatchResult::No => None,
-            });
-            query_match_futures.push(result_future);
-        }
-
-        join_all(query_match_futures).map(|results| results.into_iter().filter_map(|x| x).collect())
-    }
-
-    fn update_pending_transactions(&mut self) {
-        trace!("Updating pending matching transactions");
-        let mut pending_transactions = self.pending_transactions.lock().unwrap();
-        pending_transactions
-            .iter_mut()
-            .for_each(|utx| utx.pending_confirmations -= 1);
-
-        pending_transactions.iter().for_each(|utx| {
-            if utx.pending_confirmations == 0 {
-                let confirmed_tx_id = &utx.tx_id;
-                trace!(
-                    "Transaction {} now has enough confirmations. Sent to query result repository",
-                    confirmed_tx_id
-                );
-                self.transaction_results
-                    .add_result(utx.matching_query_id, confirmed_tx_id.clone())
+pub fn check_block_queries(
+    block_queries: ArcQueryRepository<BlockQuery>,
+    block: Block,
+) -> impl Stream<Item = QueryMatch, Error = ()> {
+    trace!("Processing {:?}", block);
+    let mut result_futures = vec![];
+
+    // We must collect the futures in a vector first to stop
+    // borrow checker freaking out
+    for (query_id, query) in block_queries.all() {
+        trace!("Matching query {:#?} against block {:#?}", query, block);
+        let block = block.clone();
+        let result_future = query.matches(&block).map(move |block_matches| {
+            if block_matches {
+                let block_id = block.as_ref().bitcoin_hash().to_string();
+                trace!("Query {:?} matches block {}", query_id, block_id);
+                Some(QueryMatch(query_id.into(), block_id))
+            } else {
+                None
             }
         });
 
-        pending_transactions.retain(|ref utx| utx.pending_confirmations > 0);
+        result_futures.push(result_future);
     }
 
-    fn process_new_transaction(
-        transaction_queries: ArcQueryRepository<TQ>,
-        pending_transactions: Arc<Mutex<Vec<PendingTransaction>>>,
-        transaction: &T,
-    ) -> impl Future<Item = Vec<QueryMatch>, Error = ()> + Send {
+    stream::futures_ordered(result_futures).filter_map(|x| x)
+}
+
+pub fn check_transaction_queries(
+    transaction_queries: ArcQueryRepository<TransactionQuery>,
+    block: Block,
+) -> impl Stream<Item = QueryMatch, Error = ()> + Send {
```

Comment:
❌ I thought the main point of this PR was to remove all this futures stuff from the bitcoin side of things?
`query.matches` should just return the result and these should just go into a vector that's returned.

Apologies if I missed some discussion.